### PR TITLE
Issue/6528 media picker insert lag

### LINKED ISSUE #6528: Media: Lag when selecting multiple images in media picker
I experience significant lag when I select multiple images in the media picker. I timed it, and after selected 15 images and tapping the checkmark, the app froze for 15 seconds before adding the images to the post content.

The lag increases relative to the number of images selected, and occurs regardless of the editor (happens in both hybrid and Aztec editors).

### Steps to reproduce the behavior

1. Start a new post.
2. Tap the Add Media button.
3. Long tap to select multiple images.
4. Tap the checkmark to add the selected images to the post.

### PR DESCRIPTION
**Pull request by nbradbury**
Fixes #6528 - reduces both the actual and apparent lag when inserting multiple images from the media picker. 

The lag was caused by saving the post after each image was inserted, and by optimizing images on the main thread (which blocked the editor from showing the inserted images until they were all inserted).

We now postpone saving until after all images are inserted, and perform optimization (and rotation, etc) in the background so the editor can show each image as it's inserted.

To test:

* In the `develop` branch create a post and add a dozen images to it, then notice the lag before they show up
* Do the same thing in this branch and notice images appear much more quickly

### CODE REVIEW

File: WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java [hunk 1]
```diff
@@ -1796,76 +1787,129 @@ private void trackAddMediaFromWPLibraryEvents(boolean isVideo, long mediaId) {
...
+            if (mediaUri == null) {
+                return false;
+            }
+
+            Activity activity = EditPostActivity.this;
+
+            if (!MediaUtils.isInMediaStore(mediaUri)
+                    && !mediaUri.toString().startsWith("/")
+                    && !mediaUri.toString().startsWith("file://")) {
+                mediaUri = MediaUtils.downloadExternalMedia(activity, mediaUri);
```
Comment (daniloercoli) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java]: Not sure we can download an external file in another thread. Ex: When I share a picture to WordPress from Google Drive the file needs to be downloaded immediately. Ref: https://github.com/wordpress-mobile/WordPress-Android/issues/5818

**Comment by daniloercoli** [comment]
On Android 5 the method `showOverlay` is called with `true` but the overlay is not always shown on the screen. To replicate this just add a picture to the editor. Then add several pictures to it.

When I tried to add several pictures to the editor, by following the steps above, I see `showOverlay` called with `false`, and immediately after  `showOverlay` called with `true` but the check `if (show && overlay.getVisibility() != View.VISIBLE)` is not true and then the overlay is not shown on the screen.

File: WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java (hunk 1 above)
Comment (nbradbury) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java]: Good catch. In what situations do we need to download external media? I'm thinking that's only necessary when choosing from the device gallery, in which case we could handle it in those situations.

Comment (nbradbury) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java]: Ignore that last comment, I'll change it to fetch from the main thread.

Comment (daniloercoli) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java]: > In what situations do we need to download external media? 

- When we share media to WordPress from an external app. EX: Google Drive -> Send to -> WordPress -> New Post.
- Sometime when Choosing from the device gallery.

To be safe we should check `MediaUtils.isInMediaStore(mediaUri)` before postponing the download to a background thread/task.

**Comment by nbradbury** [comment]
I fixed the fetch media issue in f563f53, and solved the overlay issue in a90434f. Ready for another round!

**Comment by daniloercoli** [comment]
Got a crash when exited the editor without waiting the background thread to finish is work.
```
java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.Object android.content.Context.getSystemService(java.lang.String)' on a null object reference
 at org.wordpress.android.util.DisplayUtils.getDisplayPixelSize(DisplayUtils.java:24)
 at org.wordpress.android.util.ImageUtils.getMaximumThumbnailWidthForEditor(ImageUtils.java:729)
 at org.wordpress.android.editor.AztecEditorFragment.appendMediaFile(AztecEditorFragment.java:641)
 at org.wordpress.android.ui.posts.EditPostActivity.addMediaVisualEditor(EditPostActivity.java:1911)
 at org.wordpress.android.ui.posts.EditPostActivity.access$4000(EditPostActivity.java:148)
 at org.wordpress.android.ui.posts.EditPostActivity$AddMediaListThread$2.run(EditPostActivity.java:1898)
```

**Comment by nbradbury** [comment]
> Got a crash when exited the editor without waiting the background thread to finish his work.

That was a newbie mistake on my part. I should've known I had to interrupt the thread when the activity finishes. Fixed in 36410bc.

**Comment by daniloercoli** [comment]
36410bc fixed the crash, but pictures in the thread queue aren't added to the post, if I exit the editor before it finishes his work. To replicate this just select and add a bunch of pictures, and immediately tap the back button as soon as the picker is closed.

**Comment by nbradbury** [comment]
> pictures in the thread queue aren't added to the post, if I exit the editor before it finishes his work.

In this situation, what do you recommend? I assumed it would be a rare case and when it happened the user should expect that not every photo is added, but perhaps I should prevent the user from backing out until all media are added?

**Comment by daniloercoli** [comment]
Yes, I agree with you - It should be a rare case even though on my Android 5 device 2 years old (June 2015) the lag is really noticeable when you select 4/5 photo or more.

If there is an easy way to prevent the backing out I would explore that option, otherwise we can call this PR ready to be merged.

**Comment by nbradbury** [comment]
@daniloercoli In 9b2ab31 I added a blocking progress dialog when the user adds three or more photos.

**Comment by daniloercoli** [comment]
I think this is ready to be merged. Will do that tomorrow morning, after some more testing on the old device.

**Comment by daniloercoli** [comment]
Look good! :shipit:

**Pull request merged**